src/server-core/src/lib.rs

```diff
@@ -156,52 +157,104 @@ pub async fn listen(
     Ok((handle, Box::pin(stream)))
 }
 
-/// Serves incoming TCP connections from `conns` using `server`.
-pub async fn serve<C, S>(mut conns: C, server: S)
+/// Serves incoming TCP connections from `conns` using `server`. `wait_timeout` is the time to wait
+/// after `conns` for outstanding connections to complete. Returns handles to the outstanding
+/// connections after `wait_timeout` has expired or all connections have completed.
+pub async fn serve<C, S>(mut conns: C, server: S, wait_timeout: Option<Duration>) -> JoinSet<()>
 where
     C: ConnectionStream,
     S: Server,
 {
     let task_name = format!("handle_{}_connection", S::NAME);
-    while let Some(conn) = conns.next().await {
-        let conn = match conn {
-            Ok(conn) => conn,
-            Err(err) => {
-                error!("error accepting connection: {}", err);
-                continue;
+    let mut set = JoinSet::new();
+    loop {
+        tokio::select! {
+            // TODO: Determine if next() is cancel safe. If it is not, we could drop a new
```

Comment:
I tried to figure this out but was not able. Anyone know if streams are cancel safe?